src/data/dfinsupp/basic.lean

```diff
@@ -1040,6 +1041,106 @@ assume f g, decidable_of_iff (f.support = g.support ∧ (∀i∈f.support, f i =
         by rw [hf, hg],
     by intro h; subst h; simp⟩
 
+section equiv
+open finset
+variables [Π i, has_zero (β i)]
+
+instance [is_empty ι] : unique (Π₀ i, β i) :=
+⟨⟨0⟩, λ a, by { ext, exact is_empty_elim i }⟩
+
+variables {κ : Type*} [decidable_eq κ] [Π i (x : β i), decidable (x ≠ 0)]
+/--Reindexing (and possibly removing) terms of a dfinsupp.-/
+noncomputable def congr_left (h : κ → ι) (hh : function.injective h) (f : Π₀ i, β i) :
+(Π₀ k, β (h k)) := mk (f.support.preimage h $ hh.inj_on _) (λ i, f (h i))
+
+@[simp] lemma congr_left_apply (h : κ → ι) (hh : function.injective h) (f : Π₀ i, β i) (k : κ) :
+congr_left h hh f k = f (h k) :=
+begin
+  dunfold congr_left, by_cases h0 : f (h k) = 0,
+  { rw [h0, mk_of_not_mem], rw [mem_preimage, mem_support_to_fun, not_ne_iff, h0] },
+  { apply mk_of_mem, rw [mem_preimage, mem_support_to_fun], exact h0 }
+end
+
+/--Reindexing terms of a dfinsupp.-/
+noncomputable def congr_left_equiv (h : ι ≃ κ) : (Π₀ i, β i) ≃ (Π₀ k, β (h.symm k)) :=
+⟨congr_left h.symm h.symm.injective,
+λ f, map_range (λ i, equiv.cast $ congr_arg β $ h.symm_apply_apply i)
+  (λ i, (equiv.cast_eq_iff_heq _).mpr $
+    by { convert heq.rfl, repeat { exact (h.symm_apply_apply i).symm } })
+  (@congr_left _ _ _ _ _ _ _ h h.injective f),
+λ f, by { ext i, rw [map_range_apply, congr_left_apply, congr_left_apply,
+  equiv.cast_eq_iff_heq, h.symm_apply_apply] },
+λ f, by { ext k, rw [congr_left_apply, map_range_apply, congr_left_apply,
+  equiv.cast_eq_iff_heq, h.apply_symm_apply] }⟩
+
+@[simp] lemma congr_left_equiv_apply (h : ι ≃ κ) (f : Π₀ i, β i) (k : κ) :
+congr_left_equiv h f k = f (h.symm k) := congr_left_apply h.symm h.symm.injective f k
+
+section curry
+variables {α : ι → Type*} {δ : (Σ i, α i) → Type v} [Π i, has_zero (δ i)]
+[Π i, decidable_eq (α i)] [Π i (x : δ i), decidable (x ≠ 0)]
+
+/--The natural map between `Π₀ (i : Σ i, α i), δ i` and `Π₀ i (j : α i), δ ⟨i, j⟩`.-/
+noncomputable def curry (f : Π₀ i, δ i) : Π₀ i j, δ ⟨i, j⟩ :=
+mk (f.support.image $ λ i, i.1)
+  (λ i, mk (f.support.preimage (sigma.mk i) $ sigma_mk_injective.inj_on _) $ λ j, f ⟨i, j⟩)
```

Comment:
This ought to be possible computably. I actually tried this way back in #3515, before I knew what I was doing.

Reply:
In fact, it looks like in that PR I was able to define `curry` but not `uncurry`, but your `uncurry` looks fine!

Reply:
Here's a reduced version of what I had there:
```lean
import data.dfinsupp.basic

variables {ι₁ ι₂ : Type*} (β : ι₁ → ι₂ → Type*) [Π i j, has_zero (β i j)]

def curry (f : Π₀ (ij : ι₁ × ι₂), β ij.1 ij.2) : Π₀ i, Π₀ j, β i j :=
begin
  refine quotient.lift_on f (λ fij, _) (λ a b H, _),
  exact ⟦{
    to_fun := λ i, ⟦ {
      to_fun := λ j, fij.to_fun (i, j),
      pre_support := fij.pre_support.map prod.snd,
      zero := λ j, begin
        obtain hin | h0 := fij.zero (i, j),
        {apply or.inl, rw multiset.mem_map, use (i, j), simp only [hin, eq_self_iff_true, and_self]},
        {exact or.inr h0},
      end
    } ⟧,
    pre_support := fij.pre_support.map prod.fst,
    zero := λ i, begin
      unfold has_zero.zero,
      rw quotient.eq,
      unfold has_equiv.equiv setoid.r dfinsupp.pre.to_fun,
      rw ← forall_or_distrib_left,
      intro j,
      obtain hin | h0 := fij.zero (i, j),
      { apply or.inl, rw multiset.mem_map,
        exact ⟨(i, j), hin, rfl⟩ },
      { exact or.inr h0},
    end
  }⟧,
  exact (quotient.sound $ λ i, quotient.sound $ λ j, H (i, j)),
end
```

Reply:
Actually, it turns out you was doing the case of a dfinsupp over a `ι₁ × ι₂` type while I'm doing the more general case of a sigma type `Σ i, α i`.

And at some point for `curry` you need to define the finite support of `curry f i` for all `i`, which you did by taking basically `f.support.image snd : finset ι₂`. This is bigger than the actual support of `f i` for a fixed `i : ι₁` but that's not a problem...

But in my case I need the support of `curry f i` as a `finset (α i)`, which I can't get via `f.support.image snd` as `snd ⟨i', j'⟩` is of type `α i'` and not `α i`...

Reply:
Taking `f.support.preimage (sigma.mk i) _` gives exactly the support of `curry f i`, which is thus fully in `α i` as wanted, but it's noncomputable.

Trying to use `image` instead of `preimage` requires me to provide a `(Σ i, α i) → α i`, which I tried to do : 
```
(if hi : nonempty (α i) then f.support.image (λ ⟨i', j⟩, if h : i' = i then eq.rec j h else hi.some) else ∅) : finset (α i)
```
but it's still noncomputable because of the `hi.some`.

Reply:
however `congr_left_eqquiv should be computable`, because we have an explicit inverse of the reindexation `h`.